lib/api/api.js

```diff
@@ -184,14 +186,103 @@ const api = {
             }
             return { returnTagCount, isImplicitDeny };
         }
+        let formDataParser;
+        let fileEventData = null;
+
+        if (apiMethod === 'objectPost') {
+            formDataParser = busboy({ headers: request.headers });
+        }
 
         return async.waterfall([
+            next => {
+                if (apiMethod === 'objectPut' || apiMethod === 'objectPutPart') {
+                    return next(null);
+                }
+                if (apiMethod === 'objectPost' && request.headers['content-type'].includes('multipart/form-data')) {
+                    writeContinue(request, response);
+
+                    request.formData = {};
+                    let totalFieldSize = 0;
+                    const MAX_FIELD_SIZE = 20 * 1024; // 20KB
+                    formDataParser.on('field', (fieldname, val) => {
+                        totalFieldSize += Buffer.byteLength(val, 'utf8');
+                        if (totalFieldSize > MAX_FIELD_SIZE) {
+                            const err = errors.MaxPostPreDataLengthExceeded;
+                            formDataParser.emit('error', err);
+                        }
+                        // Convert fieldname to lowercase for case-insensitive comparison
+                        const lowerFieldname = fieldname.toLowerCase();
+                        request.formData[lowerFieldname] = val;
+                    });
+
+                    formDataParser.on('file', (fieldname, file, filename, encoding, mimetype) => {
+                        fileEventData = { fieldname, file, filename, encoding, mimetype };
+                        return next(null);
+                    });
+
+                    formDataParser.on('finish', () => {
+                        // No file field == error
+                        if (!fileEventData) {
+                            const err = errors.InvalidArgument
+                                .customizeDescription('POST requires exactly one file upload per request.');
+                            formDataParser.emit('error', err);
+                        }
+                        return undefined;
+                    });
+
+                    formDataParser.on('error', (err) => {
```

Comment:
This needs to be listened earlier, otherwise an error can be uncaught and produce this bug: https://github.com/scality/Arsenal/pull/2248/files#r1669401971

Because PostObject with formData is particular, it might be better and easier to have its own function, as `callApiMethod` is already kind of a mess, we might be better off not interfering with it

Reply:
About the callback flow handling.

You could use `once` instead of `on` if you want to trigger the callback on event only once.
You can use a flag you set to true in an event callback to not execute its code again
```js
let executed = false
stream.on('file', args => {
  if (executed) {
    // ...
    return
  }
  executed = true
  // ...
  return next(null)
}
```

Also on error you could remove all listener or destroy the stream (as the doc says autoDestroy is false by default on busboy)
Also it might be better to handle your error separately than emitting them on the stream 'error' event

Reply:
On the subject of the code snippet: I thought of this approach in using a flag. However, the issue remains in making sure the API returns an error in the case the file event is fired multiple times

I suppose this is where separating the stream errors from the resource cleaning ones comes in, but I need to think how I can approach this

Reply:
This has been reworked

Reply:
see https://github.com/scality/cloudserver/pull/5601#discussion_r1690995359 I'm closing this